pyomo/dae/flatten.py

```diff
@@ -234,47 +239,108 @@ def slice_component_along_sets(
+        ComponentData is returned.
 
-    `index_map` is potentially a map from each set in `sets` to a 
-    "representative index" to use when descending into subblocks.
     """
     if type(slice_) is IndexedComponent_slice:
         context_slice = slice_.duplicate()
     else:
         context_slice = None
 
-    # Looks for components indexed by these sets immediately in our block
-    for c in b.component_objects(ctype, descend_into=False):
+    # If active argument is specified and does not match the block's
+    # active flag, we return immediately. This matches the behavior of
+    # component_objects. We only need this check as we may modify the
+    # active argument sent to component_objects if ctype is not an
+    # ActiveComponent type.
+    if active is not None and active != b.active:
+        return
+
+    # Define this class so we don't have to call issubclass again later.
+    check_active = issubclass(ctype, ActiveComponent) and (active != None)
+
+    # If active=False and ctype is not an ActiveComponent (e.g. it is Var)
+    # we will not generate any components. To prevent this, only pass the
+    # active argument if we are looking for active components.
+    c_active = active if check_active else None
+
+    # Looks for components indexed by specified sets immediately in our block.
+    for c in b.component_objects(ctype, descend_into=False, active=c_active):
         subsets = list(c.index_set().subsets())
         new_sets = [s for s in subsets if s in sets]
         # Extend our "index stack"
         sliced_sets = index_stack + new_sets
 
         # Extend our slice with this component
         for idx, new_slice in slice_component_along_sets(
-                c, sets, context_slice=context_slice, normalize=False
-                ):
-            yield sliced_sets, new_slice
+            c, sets, context_slice=context_slice, normalize=False
+        ):
+            # If we have to check activity, check data objects defined by
+            # slice. If any match, we yield the slice. This is done for
+            # compatibility with the behavior when slicing blocks, where
+            # we can only descend into a block that matches our active flag.
+            #
+            # Note that new_slice can be a data object. This happens if the
+            # component doesn't contain any sets we are slicing, i.e. new_sets
+            # is empty.
+            if (
+                # Yield if (a) we're not checking activity
+                not check_active
+                # or (b) we have not sliced and data object activity matches
+                or (not sliced_sets and new_slice.active == c_active)
+                # or (c) we did slice and *any* data object activity matches
+                or (sliced_sets and any(
+                    data.active == c_active for data in new_slice.duplicate()
+                ))
+            ):
+                yield sliced_sets, new_slice
```

Comment:
This is where we check a slice for an active data object. This is for consistency with behavior when slicing blocks in the hierarchy.